### PR TITLE
Upgrade to Fetch API + better caching

### LINKED ISSUE #1: Rely on `window.caches`
The implementation relies on the service-worker for evergreen browsers. The service-worker also caches the responses with `caches.put` etc...

It would be better to directly use the cache in the element, since the `cache` is also available there (it is not specific to the service-worker). This would remove the need for the shared [`responseCache`](https://github.com/preview-code/github-ajax/blob/a7c73751b4ec763708b591f56f7147c46aef8c07/github-ajax.html#L42), as well as the duplicate cache in the service-worker. See https://developer.mozilla.org/nl/docs/Web/API/Cache#Browser_compatibility for the compatibility of browsers and see if you find it worth it :smile:

### PR DESCRIPTION
**Pull request by txsmith**
This PR makes `github-ajax`:
- [x] Use the Fetch API
- [x] Cache stuff even without a service worker
- [x] Well tested

Fixes #1

### CODE REVIEW

File: bower.json [hunk 1]
```diff
@@ -20,8 +19,9 @@
   "devDependencies": {
     "iron-component-page": "PolymerElements/iron-component-page#^1.0.0",
     "iron-demo-helpers": "PolymerElements/iron-demo-helpers#^1.0.0",
-    "web-component-tester": "^5.0.0",
-    "webcomponentsjs": "webcomponents/webcomponentsjs#^0.7.0"
+    "web-component-tester": "^5.0.1",
+    "webcomponentsjs": "webcomponents/webcomponentsjs",
```
Comment (eanker) [inline bower.json]: Why did you remove the version number?

File: demo/basic.html [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    ></github-ajax>
+<!--
+    <github-ajax
+      id="githubSearch"
+      refurl="https://api.github.com/user"
+      last-response="{{githubUser}}"
+      on-error="_error"
+      on-request="_req"
+      on-response="_res"
+      ></github-authenticated-ajax>-->
```
Comment (eanker) [inline demo/basic.html]: Is this for future reference, or can this be deleted.

File: test/index.html [hunk 3]
```diff
@@ -18,7 +18,9 @@
   <script>
     WCT.loadSuites([
       'github-ajax.html',
-      'github-ajax.html?dom=shadow',
+      // 'github-ajax.html?dom=shadow',
+      // 'iron-ajax.html',
```
Comment (eanker) [inline test/index.html]: Shouldn't we delete all those

**Review by eanker — COMMENTED** [review]
For now I only added a few non-code related errors.
Can you also fix the linter, as it gives errors.

File: github-ajax.html [hunk 4]
```diff
@@ -4,125 +4,435 @@
...
+        ]
+
+        if (r instanceof Request) {
+          for (var i = 0, l = nonCachedContent.length; i < l; i++) {
+            if (nonCachedContent[i] === r.headers.get('Accept'))
+              return false;
+          }
+          return !this.noCache && this.method === 'GET';
+        } else if (r instanceof Response) {
+          for (var i = 0, l = nonCachedContent.length; i < l; i++) {
```
Comment (eanker) [inline github-ajax.html]: Same as above

File: github-ajax.html [hunk 5]
```diff
@@ -4,125 +4,435 @@
...
+
+      _shouldUseCache: function(r) {
+        const nonCachedContent = [
+          'application/vnd.github.VERSION.diff',
+          'application/vnd.github.black-cat-preview+json',
+          'application/vnd.github.VERSION.raw'
+        ]
+
+        if (r instanceof Request) {
+          for (var i = 0, l = nonCachedContent.length; i < l; i++) {
```
Comment (eanker) [inline github-ajax.html]: Why not 
`for (var i = 0; i < nonCachedContent.length; i++) {`

File: demo/basic.html (hunk 2 above)
Comment (txsmith) [inline demo/basic.html]: :sweat_smile:I don't even remember why we kept this

File: github-ajax.html (hunk 5 above)
Comment (txsmith) [reply github-ajax.html]: This version is faster since it does not have to access the `length` property in each iteration

**Review by eanker — APPROVED** [review]
Together we did a manual walktrough

**Comment by TimvdLippe** [comment]
Caching of all requests will not work I think? Given that the same url can be requested but different content type will therefore conflict? E.g. if you both request api info + diff and only cache api info, the diff will be broken.

**Comment by txsmith** [comment]
Yes we have noticed. I'm still looking for a fix, but meanwhile caching these things does no harm as the ETags will always differ

**Comment by TimvdLippe** [comment]
You can probably use a different cache for each content type
